File: DSAlgorithms/Leetcode/c++/383-ransomNote.cpp
#include <iostream>
#include <unordered_map>
bool canConstruct(std::string ransomNote, std::string magazine);
int main()
{
    return 0;
}
bool canConstruct(std::string ransomNote, std::string magazine)
{
    int ransomNoteSize = ransomNote.size();
    int magazineSize = magazine.size();
    std::unordered_map<char, int> hash;
    if(ransomNoteSize > magazineSize)
        return false;
    for(int character = 0; character < magazineSize; character++)
    {
        if(hash[magazine[character]])
            hash[magazine[character]]++;
        else
            hash[magazine[character]] = 1;
    }
    for(int character = 0; character < ransomNoteSize; character++)
    {
        if(hash[ransomNote[character]] && hash[ransomNote[character]] > 0)
            hash[ransomNote[character]]--;
        else
            return false;
    }
    return true;
}

File: DSAlgorithms/Leetcode/ts/src/383-ransomNote.ts
const canConstruct = (magazine: string, ransomNote: string): boolean => {
    const magazineSize = magazine.length;
    const ransomNoteSize = ransomNote.length;
    const hash = new Map<string, number>();
    if(ransomNoteSize > magazineSize)
        return false;
    for(let char = 0; char < magazineSize; char++)
    {
        if(hash.has(magazine[char]))
        {
            let count = hash.get(magazine[char]);
            if(count)
                hash.set(magazine[char], ++count);
        }
        else
            hash.set(magazine[char], 1);
    }
    for(let char = 0; char < ransomNoteSize; char++)
    {
        let count = hash.get(ransomNote[char]);
        if(count && hash.has(ransomNote[char]) && count > 0)
            hash.set(ransomNote[char], --count);
        else
            return false;
    }
    return true;
}